quinn-proto/src/connection/mod.rs

```diff
@@ -3258,78 +3234,33 @@ impl Connection {
         now: Instant,
         packet: &mut Packet,
     ) -> Result<Option<u64>, Option<TransportError>> {
-        if !packet.header.is_protected() {
-            // Unprotected packets also don't have packet numbers
-            return Ok(None);
-        }
-        let space = packet.header.space();
-        let rx_packet = self.spaces[space].rx_packet;
-        let number = packet.header.number().ok_or(None)?.expand(rx_packet + 1);
-        let key_phase = packet.header.key_phase();
-
-        let mut crypto_update = false;
-        let crypto = if packet.header.is_0rtt() {
-            &self.zero_rtt_crypto.as_ref().unwrap().packet
-        } else if key_phase == self.key_phase || space != SpaceId::Data {
-            &self.spaces[space].crypto.as_mut().unwrap().packet.remote
-        } else if let Some(prev) = self.prev_crypto.as_ref().and_then(|crypto| {
-            // If this packet comes prior to acknowledgment of the key update by the peer,
-            if crypto.end_packet.map_or(true, |(pn, _)| number < pn) {
-                // use the previous keys.
-                Some(crypto)
-            } else {
-                // Otherwise, this must be a remotely-initiated key update, so fall through to the
-                // final case.
-                None
-            }
-        }) {
-            &prev.crypto.remote
-        } else {
-            // We're in the Data space with a key phase mismatch and either there is no locally
-            // initiated key update or the locally initiated key update was acknowledged by a
-            // lower-numbered packet. The key phase mismatch must therefore represent a new
-            // remotely-initiated key update.
-            crypto_update = true;
-            &self.next_crypto.as_ref().unwrap().remote
-        };
-
-        crypto
-            .decrypt(number, &packet.header_data, &mut packet.payload)
-            .map_err(|_| {
-                trace!("decryption failed with packet number {}", number);
-                None
-            })?;
+        let result = packet_crypto::decrypt_packet_body(
```

Comment:
IMO this should also be its own separate commit.

Reply:
Just to be sure, you'd like to split https://github.com/quinn-rs/quinn/pull/1553/commits/1bf73f9d76542dd6095d04f3bf35260f4dc51108 into three commits?

1. Extracting `unprotect_header`
2. Extracting `decrypt_packet_body`
3. Adding the tests that rely on these methods

Reply:
Yes, please.